dotCMS/src/main/java/com/dotcms/rendering/velocity/viewtools/navigation/NavTool.java

```diff
@@ -73,153 +70,185 @@ protected NavResult getNav(Host host, String path) throws DotDataException, DotS
-
-                        nav.setTitle(itemFile.getFriendlyName());
+                    } else {
                         nav.setHref(ident.getURI());
-                        nav.setOrder(itemFile.getMenuOrder());
-                        nav.setType("file");
-                        nav.setPermissionId(itemFile.getPermissionId());
-                        nav.setShowOnMenu(itemFile.isShowOnMenu());
-                        children.add(nav);
                     }
+                    nav.setOrder(itemPage.getMenuOrder());
+                    nav.setType("htmlpage");
+                    nav.setPermissionId(itemPage.getPermissionId());
+                    nav.setShowOnMenu(itemPage.isShowOnMenu());
+                    children.add(nav);
                 }
-            }
+            } else if ((Contentlet.class.cast(item)).isFileAsset()) {
+                final IFileAsset itemFile = APILocator.getFileAssetAPI().fromContentlet(Contentlet.class.cast(item));
 
-            CacheLocator.getNavToolCache()
-                .putNav(host.getIdentifier(), folder.getInode(), result, languageId);
+                if ( itemFile.getLanguageId() == languageId || shouldAddFileInAnotherLang(menuItems, itemFile, languageId)) {
+                    ident = APILocator.getIdentifierAPI()
+                            .find(itemFile.getPermissionId());
+                    NavResult nav = new NavResult(folder.getInode(), host.getIdentifier(), itemFile.getLanguageId());
+
+                    nav.setTitle(itemFile.getFriendlyName());
+                    nav.setHref(ident.getURI());
+                    nav.setOrder(itemFile.getMenuOrder());
+                    nav.setType("file");
+                    nav.setPermissionId(itemFile.getPermissionId());
+                    nav.setShowOnMenu(itemFile.isShowOnMenu());
+                    children.add(nav);
+                }
+            }
+        }
+        //we only want to store things in cache for live views.
+        //It makes no sense putting things in cache when we're browsing in Preview or Edit mode
+        if(pageMode.showLive) {
+            navToolCache.putNav(host.getIdentifier(), folder.getInode(), result, languageId);
+        }
+        return new NavResultHydrated(result, this.context);
+    }
 
-            return new NavResultHydrated(result, this.context);
+    private List<?> filterNonLiveItems(final List<?> menuItems) {
+        if (!menuItems.isEmpty()) {
+            final List<Versionable> nonLive = menuItems.stream()
+                    .filter(o -> o instanceof Versionable)
+                    .map(o -> (Versionable) o)
+                    .filter(versionable -> Try.of(() -> !versionable.isLive()).getOrElse(false))
+                    .collect(Collectors.toList());
+            menuItems.removeAll(nonLive);
         }
+        return menuItems;
+    }
+
+    PageMode pageMode(){
+        final HttpServletRequest request = HttpServletRequestThreadLocal.INSTANCE.getRequest();
```

Comment:
why do you need this thing? a View tool always has the request available.

Reply:
it's a convenient method. nothing wrong there IMO